### DISCUSSION TITLE
Semantic improvement for the access modifiers.

### DISCUSSION THREAD

**Discussion by Nefcanto** [dotnet/csharplang#8236] [General]
We have these access modifiers in C#:

1. private
2. file
3. protected
4. internal
5. public

The compound word **access modifier** is very clear and understandable. It's related to **access** and it modifies how this **access**.

Now let's understand each access modifier without looking at the [docs](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/access-modifiers):

1. Private => It's probably related to a closure. Something inside curly braces. Just the way a TV could be called private if it's in my room.
2. File => If we consider a file to be the room, then access is given for the entire file. It's available in the entire file.
3. Protected => ?
4. Internal => It's related to something related to us vs. them. Then how it's different from the `private`?
5. Public => Oh, it's clear. Everyone can access it.

The point is that keywords of access modification are not self-explanatory enough. Now let's consider these terms:

1. Type => It's only accessible inside this type (class or struct)
2. File => Accessible within the same file
3. Derivation => Accessible within the derivation tree. This type or derived types at any level of derivation can access.
4. Assembly => Accessible only in the same assembly.
5. Public => World-accessible

The point is the terms used for denoting access modifiers are very vague when it comes to their relations to their true implementations and functionality.

I know after decades of existence, this change might not be possible. But just wanted to give you the feedback. Thanks.

**Comment by HaloFour** [comment]
Those names would just confuse someone else, who would have their own suggestions.  

`private`, `protected` and `public` were all taken from C++, from which C# derives.  `internal` was intentionally chosen over `assembly`, despite the latter being the name of the access modifier in IL.

**Reply by huoyaoyuan** [reply]
|IL/System.Reflection|C#|VB|System.CodeDom.MemberAttributes|
|-|-|-|-|
|public|public|Public|Public|
|family|protected|Protected|Family|
|assembly|internal|Friend|Assembly|
|private|private|Private|Private|
|famorassem|protected internal|Protected Friend|FamilyOrAssembly|
|famandassem|private protected|Private Protected|FamilyAndAssembly|

**Comment by TahirAhmadov** [comment]
I kind of like the proposed keywords, actually. Not sure about "derivation", but agree, still better than "protected".

**Reply by HaloFour** [reply]
The IL name of that accessibility modifier is `family`.
